### PR TITLE
Adding --daemon option for chef extension

### PR DESCRIPTION
**Pull request by Vasu1105**
[in progress]

### CODE REVIEW

File: lib/chef/knife/azure_server_create.rb [hunk 1]
```diff
@@ -374,7 +374,7 @@ def get_extension_status()
                   extension_status[:status] = :extension_status_not_detected
                 end
               # This fix is for linux waagent issue: api unable to deserialize the waagent status.
-              elsif role.at_css('GuestAgentStatus Status').text == 'NotReady' and waagent_status_msg == lnx_waagent_fail_msg
+              elsif role.at_css('GuestAgentStatus Status').text == 'NotReady' && waagent_status_msg == lnx_waagent_fail_msg
```
Comment (btm) [inline lib/chef/knife/azure_server_create.rb]: adding parenthesis here would be helpful for clarity, i.e.

elsif (role.at_css('GuestAgentStatus Status').text == 'NotReady') && (waagent_status_msg == lnx_waagent_fail_msg)

File: lib/chef/knife/bootstrap/bootstrap_options.rb [hunk 2]
```diff
@@ -90,6 +90,13 @@ def self.included(includer)
                                   It specifies the frequency (in minutes) at which the chef-service runs.
                                   Pass 0 if you don't want the chef-service to be installed on the target machine."
 
+            option :daemon,
+              :long => "--daemon DAEMON",
+              :description => "Optional, Provide this option when --bootstrap-protocol is set to 'cloud-api'. Works only for Windows node.
```
Comment (btm) [inline lib/chef/knife/bootstrap/bootstrap_options.rb]: "Optional. Configures the Chef client service for unattended execution. Requires --bootstrap-protocol to be 'cloud-api' and the node platform to be Windows.

File: lib/chef/knife/bootstrap/bootstrap_options.rb [hunk 3]
```diff
@@ -90,6 +90,13 @@ def self.included(includer)
...
+                                Options: 'auto' or 'service'.
+                                auto - Provide when you don't want to run the chef-client as service.
+                                service - Provide when you want to run chef-client as service"
```
Comment (btm) [inline lib/chef/knife/bootstrap/bootstrap_options.rb]: service - Configures the chef-client to run automatically in the background as a service

or "as a service or via cron" if this affects linux?

File: lib/chef/knife/bootstrap/bootstrapper.rb [hunk 4]
```diff
@@ -285,7 +285,7 @@ def get_chef_extension_public_params
           pub_config[:extendedLogs] = locate_config_value(:extended_logs) ? "true" : "false"
           pub_config[:hints] = ohai_hints if @service.instance_of?(Azure::ResourceManagement::ARMInterface) && !locate_config_value(:ohai_hints).nil?
           pub_config[:chef_service_interval] = locate_config_value(:chef_service_interval) if locate_config_value(:chef_service_interval)
-
+          pub_config[:daemon] = locate_config_value(:daemon) if locate_config_value(:daemon) && is_image_windows?
```
Comment (btm) [inline lib/chef/knife/bootstrap/bootstrapper.rb]: Should we catch here if the user isn't on windows but specified a value?

```
if locate_config_value(:daemon)
  if is_image_windows?
    pub_config[:daemon] = locate_config_value(:daemon)
  else
    raise "The daemon option is only support for Windows nodes"
  end
end
```

Comment (Vasu1105) [inline lib/chef/knife/bootstrap/bootstrapper.rb]: @btm - I am adding this in validation part instead doing this here.

**Pull request merged**